res/controllers/Reloop-Digital-Jockey-2-IE.scripts.js

```diff
@@ -0,0 +1,876 @@
+            this.inSetValue(this.isPress(channel, control, value, status));
+        } else {
+            engine.setValue(this.group, "fwd", this.isPress(channel, control, value, status));
+        }
+    },
+    unshift: function() {
+        this.inSetValue(false);     // bug? inKey was not reset by default on unshift()
+        this.inKey = "rate_temp_up";
+        this.disconnect();
+        this.outKey = "rate_temp_up";
+        this.connect();
+        this.trigger();
+    },
+    shift: function() {
+        this.inKey = "rate_temp_up_small";
+        this.disconnect();
+        this.outKey = "rate_temp_up_small";
+        this.connect();
+        this.trigger();
+    },
+});
+
+/* Jog Mode */
+
+RDJ2.JogModeSelector = function(number, searchMidiCtrl, scratchMidiCtrl, fxDryWetMidiCtrl) {
+    this.number = number;
+    this.searchMidiCtrl = searchMidiCtrl;
+    this.scratchMidiCtrl = scratchMidiCtrl;
+    this.fxDryWetMidiCtrl = fxDryWetMidiCtrl;
+    this.jogMode = RDJ2.JOGMODES.normal;
+    this.lastNonTraxJogMode = this.jogMode;
+    this.input = this.inputNormal;
+
+    components.Component.call(this);
+    this.updateControls();
+};
+RDJ2.JogModeSelector.prototype = new components.Component({
+    updateControls: function() {
+        var searchValue = this.jogMode === RDJ2.JOGMODES.search ? RDJ2.MIDI_ON : RDJ2.MIDI_OFF;
+        var scratchValue = this.jogMode === RDJ2.JOGMODES.vinyl ? RDJ2.MIDI_ON : RDJ2.MIDI_OFF;
+        var fxDryWetValue = this.jogMode === RDJ2.JOGMODES.fxdrywet ? RDJ2.MIDI_ON : RDJ2.MIDI_OFF;
+
+        if (midi.sendShortMsg) {
+            midi.sendShortMsg(0x90, this.searchMidiCtrl, searchValue);
+            midi.sendShortMsg(0x90, this.scratchMidiCtrl, scratchValue);
+            midi.sendShortMsg(0x90, this.fxDryWetMidiCtrl, fxDryWetValue);
+        }
+    },
+    inputNormal: function(channel, control, value) {
+        var isButtonPressed = RDJ2.isButtonPressed(value);
+        if (isButtonPressed) {
+            switch (control) {
+            case this.searchMidiCtrl:
+                this.jogMode = this.jogMode === RDJ2.JOGMODES.search ? RDJ2.JOGMODES.normal : RDJ2.JOGMODES.search;
+                break;
+            case this.scratchMidiCtrl:
+                this.jogMode = this.jogMode === RDJ2.JOGMODES.vinyl ? RDJ2.JOGMODES.normal : RDJ2.JOGMODES.vinyl;
+                break;
+            case this.fxDryWetMidiCtrl:
+                this.jogMode = this.jogMode === RDJ2.JOGMODES.fxdrywet ? RDJ2.JOGMODES.normal : RDJ2.JOGMODES.fxdrywet;
+                break;
+            default:
+                RDJ2.logError("Unexpected MIDI ctrl value: " + control);
+            }
+            if (this.jogMode !== RDJ2.JOGMODES.vinyl && engine.isScratching(this.number)) {
+                engine.scratchDisable(this.number, RDJ2.JOG_SCRATCH_RAMP);
+            }
+            this.updateControls();
+        }
+    },
+    inputTrax: function(channel, control, value) {
```

Comment:
since this is called `inputTrax`, I assume this has something to do with the Trax Encoder at the top of the controller? 

I don't understand the coupling between that Encoder and the jogwheel.

Reply:
When you press the Trax encoder/button it toggles Mixxx mode between "normal" mode and "library" mode. In library mode the jogs get "library browsing" functionality (turning the jogs allows choosing the track).